coreos-release-notes-generator: init script for generating release notes

zonggen opened the pull request:
This script generates Fedora CoreOS release note from pre-mature release note snippets. Takes two arguments: `--config-dir` and `--output-dir`.

The `config_dir` is the directory where a `'release-notes.d` directory resides, typically a `fedora-coreos-config` directory, which has pre-mature release notes yaml snippets.

The `output_dir` is the directory where a `release-notes.yaml` resides. If there's no such file then a new file will be created. If a `release-notes.yaml` already exists, then checks if the `next-release` field is present in the file. It inserts the `next-release` field to the front of the file if none is present or updates the existing `next-release` field if one already exists.

Related: https://github.com/coreos/fedora-coreos-tracker/issues/194
Signed-off-by: Allen Bai <abai@redhat.com>

dustymabe commented:
Thanks @zonggen for working on this! A few questions/comments:

- Can we add a nice README to this directory that explains how to use it and also what format files it expects? Some of the information from https://github.com/coreos/fedora-coreos-config/pull/550 was useful in putting the pieces together.
- Can we break up the file so that it uses functions more. I would break it up so that each logical piece is in its own function and there is a `main` function that gets run (for inspiration see https://github.com/coreos/fedora-coreos-releng-automation/blob/master/coreos-ostree-importer/coreos_ostree_importer.py#L369). Don't worry about breaking it down too much, just getting it set up that way should help it flow a bit better.

zonggen commented:
@dustymabe Thanks for reviewing! Will cleanup the code and add a README for documentation. Also will update with the updated schema in https://github.com/coreos/fedora-coreos-config/pull/550

zonggen commented:
Removed `next-release` intermediate logic and added documentations.

Currently working on adding a list of file names such that no duplication between multiple `testing` releases, since we need yaml snippets to accumulate under `release-notes.d/` for `stable`.

dustymabe commented on coreos-release-note-generator/README.md:
```diff
@@ -0,0 +1,82 @@
+# coreos-release-note-generator
```
Should we rename to `coreos-release-notes-generator`?

zonggen replied:
Agreed :+1:

dustymabe commented on coreos-release-note-generator/README.md:
```diff
@@ -0,0 +1,82 @@
+`coreos-release-note-generator` generates `release-notes.yaml` from `release-notes.d` yaml snippets using the specified build id using `--build-id` option.
+
+## Options
+ - `--build-id`: build id of the latest release note
```
```suggestion
 - `--build-id`: build id of the latest release
```

dustymabe commented on coreos-release-note-generator/release-note-generator.py:
```diff
@@ -0,0 +1,167 @@
+        args.config_dir, 'release-notes.d/*.yaml'))
+    if len(snippet_yaml_list) == 0:
+        print("release-notes.d/ does not contain any yaml snippets under '{}'".format(args.config_dir))
+        exit(0)
```
Can we return here instead of exiting in the middle of a function?

dustymabe replied:
Also, we should probably consider the case where we haven't added any release-notes.d/ files for a particular release. I guess in that case we should just put an empty entry in the list saying there are not release notes or something.

zonggen replied:
Or, we could just merge this *list* of *build_id: release_notes* items into a single dictionary so that we could query whether a release note exists for a release just by checking if the key *build_id* exists. This could be done when we are converting to JSON format for website consumption, on the other hand I think it makes sense to use YAML file for human readability on other places like `fedora-coreos-config`.

dustymabe commented on coreos-release-note-generator/release-note-generator.py:
```diff
@@ -0,0 +1,167 @@
+                note = {'subject': item.get(
+                    'subject', ''), 'body': item.get('body', '')}
+                # purposely avoid item.get('component', '') to error out if the component key does not exist
+                project_name = item['component']
```
instead of using `project` throughout this file should we just call it `component` instead?

zonggen replied:
Agreed :+1:

dustymabe commented:
one small nit and then we can merge:

Change commit subject line to `coreos-release-notes-generator: init script for generating release notes`

zonggen commented:
Fixed commit subject, thanks for pointing that out!

dustymabe commented:
LGTM - I think we can merge this and do fixups for things we find in a followup PR.
